src/audio/crossover/crossover.c

```diff
@@ -0,0 +1,745 @@
+#include <sof/common.h>
+#include <sof/debug/panic.h>
+#include <sof/drivers/ipc.h>
+#include <sof/lib/alloc.h>
+#include <sof/lib/memory.h>
+#include <sof/lib/uuid.h>
+#include <sof/list.h>
+#include <sof/platform.h>
+#include <sof/string.h>
+#include <sof/trace/trace.h>
+#include <ipc/control.h>
+#include <ipc/stream.h>
+#include <ipc/topology.h>
+#include <user/trace.h>
+#include <user/crossover.h>
+#include <errno.h>
+#include <stddef.h>
+#include <stdint.h>
+
+static const struct comp_driver comp_crossover;
+
+/* 948c9ad1-806a-4131-ad6c-b2bda9e35a9f */
+DECLARE_SOF_UUID("crossover", crossover_uuid, 0x948c9ad1, 0x806a, 0x4131,
+		 0xad, 0x6c, 0xb2, 0xbd, 0xa9, 0xe3, 0x5a, 0x9f);
+
+/* FOR TESTING ONLY - REMOVE THIS !!!!!!!! */
+int lp0[] = {0xc253826f, 0x7da1773e, 0x000161a1, 0x0002c342, 0x000161a1, 0, 16484};
+int lp1[] = {0xcad0cdef, 0x742e8c5d, 0x00202837, 0x0040506d, 0x00202837, 0, 16484};
+int lp2[] = {0xe16f20ea, 0x51e57f66, 0x01966267, 0x032cc4ce, 0x01966267, 0, 16484};
+
+int hp0[] = {0xc253826f, 0x7da1773e, 0x1f7cd6e3, 0xc106523b, 0x1f7cd6e3, 0, 16484};
+int hp1[] = {0xcad0cdef, 0x742e8c5d, 0x1d3d7328, 0xc58519af, 0x1d3d7328, 0, 16484};
+int hp2[] = {0xe16f20ea, 0x51e57f66, 0x161c344b, 0xd3c7976a, 0x161c344b, 0, 16484};
+
+int *lp[] = {lp0, lp1, lp2};
+int *hp[] = {hp0, hp1, hp2};
+/* DO NOT FORGET TO REMOVE */
+
+static void crossover_free_config(struct sof_crossover_config **config)
+{
+		rfree(*config);
+		*config = NULL;
+}
+
+static void crossover_reset_state_lr4(lr4_state *lr4)
+{
+	rfree(lr4->delay);
+
+	lr4->coef = NULL;
+	lr4->delay = NULL;
+}
+
+static void crossover_reset_state_ch(struct crossover_state *ch_state)
+{
+	int i;
+
+	for (i = 0; i < 3; i++) {
+		crossover_reset_state_lr4(&ch_state->lowpass[i]);
+		crossover_reset_state_lr4(&ch_state->highpass[i]);
+	}
+}
+
+static void crossover_reset_state(struct comp_data *cd)
+{
+	int i;
+
+	for (i = 0; i < PLATFORM_MAX_CHANNELS; i++)
+		crossover_reset_state_ch(&cd->state[i]);
+}
+
+/**
+ * The assign_sink array in the configuration maps to the pipeline ids.
+ * This function  returns the index i such that assign_sink[i] = pipe_id
+ */
+static uint8_t get_stream_index(struct sof_crossover_config *config,
+				uint32_t pipe_id)
+{
+	int i;
+	int32_t *assign_sink = &config->data[0];
+
+	for (i = 0; i < config->num_sinks; i++)
+		if (assign_sink[i] == pipe_id)
+			return i;
+
+	comp_cl_err(&comp_crossover, "get_stream_index() error: couldn't find configuration for connected pipeline %u",
+		    pipe_id);
+
+	return -EINVAL;
+}
+
+static int crossover_assign_sinks(struct comp_dev *dev,
+				  struct sof_crossover_config *config,
+				  struct comp_buffer **sinks)
+{
+	struct comp_buffer *sink;
+	struct list_item *clist;
+	int num_sinks = 0;
+	int i;
+
+	// align sink streams with their respective configurations
+	list_for_item(clist, &dev->bsink_list) {
+		sink = container_of(clist, struct comp_buffer, source_list);
+		if (sink->sink->state == dev->state) {
```

Comment:
@cujomalainey If a sink is not ready then even if the config is valid, we may get a gap in the array of sinks. Do you know if this scenario is possible?

Reply:
I tried creating this topology (inspired from the demux topologies):
PCM -> Crossover -> DAI
                    |
                    pipeline 2: -> Virtual DAI

And it seems that it is possible that one of the buffers is in a different state from the component.